module/os/freebsd/zfs/zfs_vnops_os.c

```diff
@@ -2920,6 +2920,66 @@ zfs_setattr(znode_t *zp, vattr_t *vap, int flags, cred_t *cr)
 	return (err);
 }
 
+/*
+ * Look up the directory entries corresponding to the source and target
+ * directory/name pairs.
+ */
+static int
+zfs_rename_relock_lookup(znode_t *sdzp, const struct componentname *scnp,
+    znode_t **szpp, znode_t *tdzp, const struct componentname *tcnp,
+    znode_t **tzpp)
+{
+	zfsvfs_t *zfsvfs;
+	znode_t *szp, *tzp;
+	int error;
+
+	/*
+	 * Before using sdzp and tdzp we must ensure that they are live.
+	 * As a porting legacy from illumos we have two things to worry
+	 * about.  One is typical for FreeBSD and it is that the vnode is
+	 * not reclaimed (doomed).  The other is that the znode is live.
+	 * The current code can invalidate the znode without acquiring the
+	 * corresponding vnode lock if the object represented by the znode
+	 * and vnode is no longer valid after a rollback or receive operation.
+	 * z_teardown_lock hidden behind ZFS_ENTER and ZFS_EXIT is the lock
+	 * that protects the znodes from the invalidation.
+	 */
+	zfsvfs = sdzp->z_zfsvfs;
+	ASSERT3P(zfsvfs, ==, tdzp->z_zfsvfs);
+	ZFS_ENTER(zfsvfs);
+	ZFS_VERIFY_ZP(sdzp);
```

Comment:
I believe we need to check for z_sa_hdl == NULL as before since ZFS_VERIFY_ZP() can directly return thus the code won't do ZFS_EXIT(zfsvfs).

Reply:
From my reading, the ZFS_VERIFY_ZP() will always do a ZFS_EXIT() before returning. In particular, ZFS_VERIFY_ZP() must only be used in a ZFS_ENTER/EXIT-bracketed section.

Reply:
I stand corrected and should have check the definition
```/* Verifies the znode is valid */
#define ZFS_VERIFY_ZP(zp) \
    if (__predict_false((zp)->z_sa_hdl == NULL)) { \
        ZFS_EXIT((zp)->z_zfsvfs); \
        return (EIO); \
    } \